src/librustc/middle/reachable.rs

```diff
@@ -32,9 +32,15 @@ fn item_might_be_inlined(tcx: TyCtxt<'tcx>, item: &hir::Item, attrs: CodegenFnAt
         return true
     }
 
-    match item.node {
-        hir::ItemKind::Impl(..) |
-        hir::ItemKind::Fn(..) => {
+        match item.node {
+        hir::ItemKind::Fn(_, header, ..) => {
+            if header.constness == hir::Constness::Const {
```

Comment:
Imo this should be turned into a guard condition instead (or better yet, pattern match on `constness`).

Reply:
Added a match guard.